app/src/main/java/com/kickstarter/viewmodels/MessagesViewModel.java

```diff
@@ -75,129 +103,211 @@ public ViewModel(final @NonNull Environment environment) {
+          );
+        });
+
+      final Observable<KoalaContext.Message> koalaContext = intent()
+        .map(i -> i.getSerializableExtra(IntentKey.KOALA_CONTEXT))
+        .ofType(KoalaContext.Message.class);
+
+      final Observable<Backing> configBacking = configData
+        .map(Either::right)
+        .filter(ObjectUtils::isNotNull)
+        .map(PairUtils::second);
+
+      final Observable<MessageThread> configThread = configData
+        .map(Either::left)
+        .filter(ObjectUtils::isNotNull);
+
+      final Observable<Either<Backing, MessageThread>> backingOrThread = Observable.merge(
+        configBacking.map(backing -> new Either.Left<>(backing)),
+        configThread.map(thread -> new Either.Right<>(thread))
+      );
 
-      final Observable<Notification<Message>> messageNotification = messageThread
+      final Observable<Notification<Message>> messageNotification = backingOrThread
         .compose(combineLatestPair(this.messageEditTextChanged))
         .compose(takeWhen(this.sendMessageButtonClicked))
-        .switchMap(threadAndBody ->
-          this.client.sendMessageToThread(threadAndBody.first, threadAndBody.second)
-            .materialize()
+        .switchMap(backingOrThreadAndBody ->
+          backingOrThreadAndBody.first.either(
+            backing -> this.client.sendMessageToBacking(backing, backingOrThreadAndBody.second),
+            thread -> this.client.sendMessageToThread(thread, backingOrThreadAndBody.second)
+          )
         )
+        .materialize()
         .share();
 
-      final Observable<Message> messageSent = messageNotification.compose(values());
+      final Observable<Message> messageSent = messageNotification.compose(values()).ofType(Message.class);
 
       this.setMessageEditText = messageSent.map(__ -> "");
 
-      final Observable<Notification<MessageThreadEnvelope>> envelopeNotification = Observable.merge(
-        messageThread,
-        messageThread.compose(takeWhen(messageSent))
+      final Observable<MessageThreadEnvelope> messageThreadEnvelope = Observable.merge(
+        backingOrThread,
+        backingOrThread.compose(takeWhen(messageSent))
       )
-        .switchMap(thread -> this.client.fetchMessagesForThread(thread).materialize())
+        .switchMap(bOrT -> bOrT.either(this.client::fetchMessagesForBacking, this.client::fetchMessagesForThread))
+        .compose(neverError())
         .share();
 
-      final Observable<MessageThreadEnvelope> messageThreadEnvelope = envelopeNotification.compose(values());
+      final Observable<Project> project = configData
+        .map(data -> data.either(MessageThread::project, projectAndBacking -> projectAndBacking.first));
 
-      final Observable<Notification<Backing>> backingNotification = Observable.combineLatest(
-        messageThread,
-        this.currentUser.observable(),
-        Pair::create
+      final Observable<User> participant = Observable.merge(
```

Comment:
this logic grabs the project's creator if the VM was not initialized with a MessageThread...I'm not sure if this logic is a little too buried let me know if there are ways to improve!